runtime/common/src/precompile/liquid_crowdloan.rs

```diff
@@ -0,0 +1,290 @@
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: Output::encode_uint(redeem_amount),
+					logs: Default::default(),
+				})
+			}
+			Action::GetRedeemCurrency => {
+				let currency_id = <module_liquid_crowdloan::Pallet<Runtime>>::redeem_currency();
+				let address = <Runtime as module_prices::Config>::Erc20InfoMapping::encode_evm_address(currency_id)
+					.ok_or_else(|| PrecompileFailure::Revert {
+						exit_status: ExitRevert::Reverted,
+						output: "LiquidCrowdloan redeem currency not found".into(),
+						cost: target_gas_limit(target_gas).unwrap_or_default(),
+					})?;
+
+				log::debug!(target: "evm", "liuqid_crowdloan: GetRedeemCurrency output: {:?}", currency_id);
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: Output::encode_address(address),
+					logs: Default::default(),
+				})
+			}
+		}
+	}
+}
+
+struct Pricer<R>(PhantomData<R>);
+
+impl<Runtime> Pricer<Runtime>
+where
+	Runtime: module_evm::Config + module_prices::Config + module_liquid_crowdloan::Config,
+{
+	const BASE_COST: u64 = 200;
+
+	fn cost(
+		input: &Input<Action, Runtime::AccountId, Runtime::AddressMapping, Runtime::Erc20InfoMapping>,
+	) -> Result<u64, PrecompileFailure> {
+		let action = input.action()?;
+
+		let cost = match action {
+			Action::Redeem => {
+				let read_account = InputPricer::<Runtime>::read_accounts(1);
+				let weight = <Runtime as module_liquid_crowdloan::Config>::WeightInfo::redeem();
+
+				Self::BASE_COST
+					.saturating_add(read_account)
+					.saturating_add(WeightToGas::convert(weight))
+			}
+			Action::GetRedeemCurrency => {
+				let weight = <Runtime as frame_system::Config>::DbWeight::get().reads(1);
+
+				Self::BASE_COST.saturating_add(WeightToGas::convert(weight))
+			}
+		};
+		Ok(Self::BASE_COST.saturating_add(cost))
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::precompile::mock::{
+		bob, bob_evm_addr, new_test_ext, Currencies, LiquidCrowdloan, LiquidCrowdloanPalletId, RuntimeOrigin, Test,
+		DOT, LCDOT, LDOT,
+	};
+	use frame_support::assert_ok;
+	use hex_literal::hex;
+	use orml_traits::MultiCurrency;
+	use sp_runtime::traits::AccountIdConversion;
+
+	type LiquidCrowdloanPrecompile = crate::precompile::LiquidCrowdloanPrecompile<Test>;
+
+	#[test]
+	fn redeem_dot() {
+		new_test_ext().execute_with(|| {
+			let context = Context {
+				address: Default::default(),
+				caller: bob_evm_addr(),
+				apparent_value: Default::default(),
+			};
+
+			assert_ok!(Currencies::update_balance(
+				RuntimeOrigin::root(),
+				bob(),
+				LCDOT,
+				1_000_000_000
+			));
+
+			assert_ok!(Currencies::update_balance(
+				RuntimeOrigin::root(),
+				LiquidCrowdloanPalletId::get().into_account_truncating(),
+				DOT,
+				1_000_000_000
+			));
+
+			// redeem(address,uint256) -> 1e9a6950
+			// who
+			// amount 1e9
+			let input = hex!(
```

Comment:
```suggestion
			let input = hex! { "
```
to make fmt ignore this